### PR TITLE
Txpool v2 features for futures connections with other modules

### PR DESCRIPTION
**Pull request by AurelienFT**
REVIEWER NOTE: Review https://github.com/FuelLabs/fuel-core/pull/2193 first for context
## Linked Issues/PRs
Closes https://github.com/FuelLabs/fuel-core/issues/2187

## Description

(2k of additions is the tests from TxPool v1 which has been instantiated again)

This PR adds all the logic for the service and shared state of the TxPool to be able to be used by all the others modules and to make all the auxiliary mechanisms working.

The PR changes the way the `SharedState` interact with the pool and the service. Now the `SharedState` only sends commands to the service that will answer him whenever he is time to do depending on the priority of the request (managed through a tokio select biaised). This makes all methods to return async even `find` which I'm not a big fan of but didn't found any better solution. 
**EDIT: I have made some tests especially tries to run `execute_suite` tests and it seems that blocking can cause "double blocking forever" and so I'm up for an alternative**

Here is a list of all interactions made by the different modules and their new name/behavior/reason : 

### Block producer
| Existing | New |
| - | - |
| select_transactions(max_gas) | select_transactions(max_gas) |

### PoA
| Existing | New |
| - | - |
| total_consumable_gas() | Nothing (unused) |
| transaction_status_events() | Replaced by new_txs_notifier [reason](https://github.com/FuelLabs/fuel-core/blob/d5ba082264e1e3db54c6c81f6036eb06b9d3044c/crates/services/consensus_module/poa/src/service.rs#L590) |
| pending_number() | Not used anymore after transaction_status_events() change |
| remove_txs() | broadcast_txs_skipped_reason() |

### Block Importer
| Existing | New |
| - | - |
| Stream with blockEvents | Stream with block events |

### P2P
| Existing | New |
| - | - |
| gossiped_tx_stream | tx_from_p2p_stream |
| new_tx_gossip_subscription | new_peers_subscribed_stream |
| broadcast_transaction | broadcast_transaction |
| get_tx_ids| get_tx_ids |
| get_txs | get_txs |
| notify_gossip_transaction_validity | notify_gossip_transaction_validity |

### GraphQL API

| Existing | New |
| - | - |
| transaction | transaction |
| tx_update_subscribe | tx_update_subscribe |

### Others changes

- The PR also changes the error type to help separate subdomains of errors and try to be more clear about the moment errors can happens.
- The TTL management of the transactions is also in place as the less priority task for the service.
- A new thread pool has been added to manage P2P sync that could take a long time (same behavior as txpool v1)
- The transactions are now stored as Arc because they need to be shareable with others modules (API, P2P)
- All the tests from the previous pool has been instantiated and pass. 
- The TxPool don't have anymore unused authorization and match the correct linting requirements.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### CODE REVIEW

**Comment by AurelienFT** [comment]
Making this draft given the conflicts.

**Comment by AurelienFT** [comment]
Readyf ro review again

**Comment by AurelienFT** [comment]
Change this to draft to add gas price and manage delete of skipped transactions dependents

**Comment by AurelienFT** [comment]
@xgreenx I have added the check of gas price in selection and the remove of coin dependents on skipped txs.

**Review by xgreenx — APPROVED** [review]

File: crates/services/txpool_v2/src/service.rs
Comment (acerone85) [inline crates/services/txpool_v2/src/service.rs:141]: is there any reason why you use the `Shared` type alias for pool and not for current_height? 
 Overall I think `Arc<RwLock<_>>` is more clear

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [inline crates/services/txpool_v2/src/service.rs:136]: Do we need to hold a trait object here? Can't we use a generic and save a heap allocation?

```suggestion
pub struct Task<View. P2PRequests> {
    chain_id: ChainId,
    subscriptions: Subscriptions,
    verification: Verification<View>,
    p2p: P2PRequests,
```

Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:136]: The choice has been made to reduce the generics and use the Arc even if it's less performant because they are mainly used in tasks that consume a lot and so they will not be our bottleneck. + we win code readability

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [inline crates/services/txpool_v2/src/service.rs:99]: While this is a pretty nice type-alias, I don't see any point in making it public. Any other modules that would benefit from the type alias can redefine it, which would be more straight-forward to read and just as concise as importing it without the extra indirection.

```suggestion
type Shared<T> = Arc<RwLock<T>>;
```

File: crates/services/txpool_v2/src/service/verifications.rs
Comment (netrome) [inline crates/services/txpool_v2/src/service/verifications.rs:11]: For example here I would find it more readable if we instead of importing `Shared` redefine the alias in this module.

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [reply crates/services/txpool_v2/src/service.rs:99]: Optionally if this is used in a lot of places we should move it out to another module, since the alias has nothing to do with this particular service.

**Review by netrome — COMMENTED** [review]
Some opinions on the `Shared` type, continuing my review...

File: crates/services/src/lib.rs
Comment (AurelienFT) [reply crates/services/src/lib.rs:87]: @xgreenx

File: crates/services/txpool_v2/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:99]: I changed it to pub(crate) because it's used in the tests

File: crates/services/txpool_v2/src/service/verifications.rs
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service/verifications.rs:11]: will let @xgreenx  give his advice on this as it's a type he added

File: crates/services/src/lib.rs
Comment (xgreenx) [reply crates/services/src/lib.rs:87]: image


It was done to enable feature by default during testing

File: crates/services/txpool_v2/src/service/verifications.rs
Comment (xgreenx) [reply crates/services/txpool_v2/src/service/verifications.rs:11]: Up to @AurelienFT )

Comment (AurelienFT) [reply crates/services/txpool_v2/src/service/verifications.rs:11]: Noted on my note will think about it in the next days to not block the merge.

**Review by acerone85 — APPROVED** [review]

**Review by xgreenx — APPROVED** [review]

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [reply crates/services/txpool_v2/src/service.rs:136]: Makes sense :+1:

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [inline crates/services/txpool_v2/src/service.rs:190]: I know the pattern was established before this PR, but it would make more sense to me to just return here instead of declaring and assigning the variable to later return it.

```suggestion
                    return Ok(true);
                } else {
                    return Ok(false);
```

(and the same for all other branches)

**Review by netrome — COMMENTED** [review]
Another suggestion (not blocking merge)

File: crates/services/txpool_v2/src/service.rs
Comment (netrome) [reply crates/services/txpool_v2/src/service.rs:190]: Nevermind, I see you already merged! 🎉

Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:190]: It's Ok you can leave comment and I will address in patch in the future. I note this and let this open in the meantime

File: crates/services/txpool_v2/src/service.rs
Comment (rafal-ch) [inline crates/services/txpool_v2/src/service.rs:622]: These trait bounds look too restrictive, please check if the `Send + Sync` are needed. Also unsure about `'static`.

File: crates/services/txpool_v2/src/pool.rs
Comment (rafal-ch) [inline crates/services/txpool_v2/src/pool.rs:286]: Just noticed that all tests in `crates/services/txpool_v2/src/tests/mod.rs` are still green if I remove these lines. Seems like tests don't ensure that we promote dependents to executables.

Maybe there are tests for that in different place.

File: crates/services/txpool_v2/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool_v2/src/service.rs:622]: Static is necessary on the persistent storage because we create an Arc from this. Removed the one from PSProvider.
For the `Sync` and `Send` they were added in `ports.rs` and don't removed from here. I have removed them from here, ty :) They all have `Send` `Sync` and `'static` in `ports.rs` to be usable in arcs and in the future that is defined in `async fn run`.

Changes are available here : https://github.com/FuelLabs/fuel-core/pull/2263/commits/1c345c0bc7799e8e64cece36261f105d896323af

File: crates/services/txpool_v2/src/pool.rs
Comment (AurelienFT) [reply crates/services/txpool_v2/src/pool.rs:260]: We are ok to try to delete the others if they are not already been removed yet. I'm not sure about the warning because this function can be called from a lot of differents place/modules around the code and even if it's suboptimal to call again this function (with an other module for example) on already deleted functions I don't think it's worth warning to the user about this has it will not make his node gone badly or break anything.

Comment (rafal-ch) [reply crates/services/txpool_v2/src/pool.rs:260]: If this doesn't break anything there's no need to log, indeed.

Thanks for explanation! :+1:

**Pull request merged**
